### PR TITLE
fix: redirect after login not working correctly with 401 on cms requests

### LINKED ISSUE #10977: Redirects after login not working correctly with 401 on cms requests.
### Problem

When `access_token` expires before navigation first request that will fail will be request for cms page.
In this case after user will complete login, he won't be redirected to expected page. It won't be registered by guards, because for them to kick in we need to first load cms page.

### Steps to reproduce

1. Login
2. Go to any page eg. address_book
3. Temper with your access_token in local storage
4. Refresh page
5. Login
6. Redirect will happen to homepage instead of previous page (eg. address_book)

### PR DESCRIPTION
**Pull request by Platonn**
When we make the OCC call for cms page (triggered at the beginning of `CmsPageGuard`), the call is intercepted by the `AuthInterceptor`. The interceptor's service might kick out the user to the `/login` page (in case of expired `refresh_token`) even before the back-redirect url was saved by the cms component guard (`AuthGuard`). This caused a bug of back-redirecting after login not to the last visited (or last anticipated page), but to the homepage.

Now we fix it, by:
- remembering automatically the last visited page (on every NavigationEnd routing event)
- and remembering (imperatively) the anticipated page. It's invoked by `AuthGuard` or by the `AuthHttpHeaderService` 

By the way:
- deprecated methods of `AuthRedirectService`:
   - `#reportNotAuthGuard` - method not needed anymore. Every visited URL is now remembered automatically as redirect URL on NavigationEnd event. 
   - `#reportAuthGuard` - use the new equivalent method instead: `saveCurrentNavigationUrl`

closes #10977

### CODE REVIEW

**Comment by AbdelhalimOuasti** [comment]
Hi any idea when it would be available ?

**Comment by Platonn** [comment]
@AbdelhalimOuasti, due to breaking changes it's planned for 4.0. However, no specific date is known yet for 4.0

File: projects/core/src/auth/user-auth/services/auth-http-header.service.ts [hunk 1]
```diff
@@ -95,7 +97,18 @@ export class AuthHttpHeaderService {
...
+
+    // There might be 2 cases:
+    // 1. when user is already on some page (router is stable) and performs an UI action
+    // that triggers http call (i.e. button click to save data in backend)
+    // 2. when user is navigating to some page and a route guard triggers the http call
+    // (i.e. guard loading cms page data)
+    //
+    // In the second case, we want to remember the anticipated url before we navigate to
+    // the login page, so we can redirect back to that URL after user authenticates.
+    this.authRedirectService.saveCurrentNavigationUrl();
```
Comment (marlass) [inline projects/core/src/auth/user-auth/services/auth-http-header.service.ts:105]: I'm thinking if it is safe to do it between logout and redirect? Is there any reason for not doing that before logout? I'm thinking for a case, where some customer implementation would have `coreLogout` working synchronously and making a redirect there.

File: projects/core/src/auth/user-auth/services/auth-redirect.service.spec.ts [hunk 2]
```diff
@@ -30,94 +37,118 @@ describe('AuthRedirectService', () => {
...
+      service.redirect();
+      expect(authRedirectStorageService.getRedirectUrl).toHaveBeenCalled();
+      expect(routingService.go).toHaveBeenCalledWith('/');
+    });
 
-      it('should redirect to the previous url', () => {
-        expect(routingService.goByUrl).toHaveBeenCalledWith('/test');
-      });
+    it('should clear saved redirect URL', () => {
+      authRedirectStorageService.setRedirectUrl(undefined);
```
Comment (marlass) [inline projects/core/src/auth/user-auth/services/auth-redirect.service.spec.ts]:
```suggestion
```

**Review by marlass — DISMISSED** [review]

**Review by marlass — DISMISSED** [review]

**Review by dunqan — DISMISSED** [review]

**Review by dunqan — APPROVED** [review]

**Pull request merged**